File: src/Ycs/Contracts/IUpdateDecoder.cs
// ------------------------------------------------------------------------------
//  <copyright company="Microsoft Corporation">
//      Copyright (c) Microsoft Corporation.  All rights reserved.
//  </copyright>
// ------------------------------------------------------------------------------

using System;
using System.IO;

namespace Ycs.Contracts
{
    public interface IDSDecoder : IDisposable
    {
        Stream Reader { get; }

        void ResetDsCurVal();
        long ReadDsClock();
        long ReadDsLength();
    }

    public interface IUpdateDecoder : IDSDecoder
    {
        StructID ReadLeftId();
        StructID ReadRightId();
        long ReadClient();
        byte ReadInfo();
        string ReadString();
        bool ReadParentInfo();
        uint ReadTypeRef();
        int ReadLength();
        object ReadAny();
        byte[] ReadBuffer();
        string ReadKey();
        object ReadJson();
    }
}

File: ycs-golang/structs/IUpdateDecoder.go
package structs

import "io"

// IDSDecoder is an interface that extends io.Closer (for Dispose functionality)
// and provides methods for decoding data structures.
type IDSDecoder interface {
	io.Closer
	Reader() io.Reader
	ResetDsCurVal()
	ReadDsClock() int64
	ReadDsLength() int64
}

// IUpdateDecoder extends IDSDecoder with additional methods for update decoding.
type IUpdateDecoder interface {
	IDSDecoder
	ReadLeftId() ID
	ReadRightId() ID
	ReadClient() int64
	ReadInfo() byte
	ReadString() string
	ReadParentInfo() bool
	ReadTypeRef() uint32
	ReadLength() int
	ReadAny() interface{}
	ReadBuffer() []byte
	ReadKey() string
	ReadJson() interface{}
}
